### PR TITLE
Implement DataAPI

### PR DESCRIPTION
**Pull request by quinnj**

### CODE REVIEW

File: src/fallbacks.jl [hunk 1]
```diff
@@ -59,20 +59,18 @@ function rows(x::T) where {T}
...
 # build columns from rows
-haslength(L) = L isa Union{Base.HasShape, Base.HasLength}
-
 """
     Tables.allocatecolumn(::Type{T}, len) => returns a column type (usually AbstractVector) w/ size to hold `len` elements
 
     Custom column types can override with an appropriate "scalar" element type that should dispatch to their column allocator.
 """
-allocatecolumn(T, len) = Vector{T}(undef, len)
+allocatecolumn(T, len) = DataAPI.defaultarray(T, 1)(undef, len)
```
Comment (nalimilan) [inline src/fallbacks.jl:74]: Should `defaultarray` replace `allocatecolumn` at some point?

File: src/tofromdatavalues.jl [hunk 2]
```diff
@@ -1,24 +1,21 @@
 # utilities for converting Row iterators to and from DataValue-based NamedTuple iterators
 # the DataValue-specific definitions are in the Tables.jl __init__ DataValues @require block
-nondatavaluetype(::Type{T}) where {T} = T
-nondatavaluetype(::Type{Union{}}) = Union{}
-datavaluetype(::Type{T}) where {T} = T
-datavaluetype(::Type{Union{}}) = Union{}
-
-Base.@pure function nondatavaluetype(::Type{NT}) where {NT <: NamedTuple{names}} where {names}
-    TT = Tuple{Any[ nondatavaluetype(fieldtype(NT, i)) for i = 1:fieldcount(NT) ]...}
+DataAPI.nondatavaluetype(::Type{T}) where {T} = T
```
Comment (nalimilan) [inline src/tofromdatavalues.jl]: I know that this function is documented to be owned by Tables.jl, but given how simple this fallback definition is, why not put it in DataAPI instead? That would be simpler for everybody.

File: src/fallbacks.jl (hunk 1 above)
Comment (quinnj) [reply src/fallbacks.jl:74]: they're slightly different; `defaultarray` just returns the array type for a given scalar type; allocatecolumn is for actually creating a column. We can probably just replace uses of `allocatecolumn` with `defaultarray(T)(undef, len)`

File: src/tofromdatavalues.jl (hunk 2 above)
Comment (quinnj) [reply src/tofromdatavalues.jl]: That's fair.

File: src/fallbacks.jl (hunk 1 above)
Comment (nalimilan) [reply src/fallbacks.jl:74]: Yes, that's what I had in mind.

**Pull request merged**